File: apps/api_py/src/linkbook/agents/time_sentinel.py
"""§5.3 — Time Sentinel. Proposes a self-nudge when < 4h logged."""

from __future__ import annotations

from typing import Any

from sqlalchemy.orm import Session

from ..config import AppConfig
from ..db.models import Action
from ..idempotency import compute_idempotency_key, find_duplicate_in_window
from ..lib.log import AppLogger
from ..types import ACTION_CATALOG
from .runtime import ProposeRequest, get_agentspan_client, propose_with_fallback

VERSION = "v1.0"


async def run_time_sentinel(
    event: dict[str, Any], cfg: AppConfig, db: Session, log: AppLogger
) -> None:
    if event["type"] != "time.missing_yesterday":
        return

    client = get_agentspan_client(cfg)
    result = await propose_with_fallback(
        client,
        ProposeRequest(
            agent="time_sentinel", agent_version=VERSION, event=event, context={}
        ),
        log,
    )
    if not result.ok:
        return

    p = event["payload"]
    msg = (
        f"You logged {p['hours_logged']}h on {p['date']}. "
        "Mind finishing your timesheet before Friday?"
    )
    params = {"message": msg}
    idem = compute_idempotency_key("time.self_nudge", event["subject_ref"], params)
    dup = find_duplicate_in_window(db, idem)
    if dup is not None:
        log.info({"idem": idem, "prior": dup.id}, "time sentinel: duplicate within 24h, skipping")
        return

    db.add(
        Action(
            type="time.self_nudge",
            params=params,
            mode="proposed",
            drafted_by=f"agent:time_sentinel@{VERSION}",
            status="drafted",
            reversal_class=ACTION_CATALOG["time.self_nudge"].reversal_class,
            idempotency_key=idem,
            agent_confidence=str(result.proposal.confidence),
            agent_rationale=result.proposal.rationale,
            preview=msg,
            originating_event_id=event["id"],
            subject_ref=event["subject_ref"],
        )
    )
    db.commit()

File: apps/api/src/agents/time-sentinel.ts
import type { Event } from '@linkbook/types';
import { actionCatalog } from '@linkbook/types';
import { schema } from '@linkbook/db';
import type { AppCtx } from '../lib/principal-context.js';
import { getAgentspanClient, proposeWithFallback } from './runtime.js';
import { computeIdempotencyKey, findDuplicateInWindow } from '../idempotency.js';

const VERSION = 'v1.0';

export async function runTimeSentinel(event: Event, ctx: AppCtx): Promise<void> {
  if (event.type !== 'time.missing_yesterday') return;

  const client = getAgentspanClient(ctx.cfg);
  const result = await proposeWithFallback(
    client,
    { agent: 'time_sentinel', agent_version: VERSION, event, context: {} },
    ctx.log,
  );
  if (!result.ok) return;

  const params = {
    message: `You logged ${event.payload.hours_logged}h on ${event.payload.date}. Mind finishing your timesheet before Friday?`,
  };
  const idem = computeIdempotencyKey('time.self_nudge', event.subject_ref, params);
  const dup = await findDuplicateInWindow(ctx, idem);
  if (dup) {
    ctx.log.info({ idem, prior: dup.id }, 'time sentinel: duplicate within 24h, skipping');
    return;
  }

  ctx.db
    .insert(schema.actions)
    .values({
      type: 'time.self_nudge',
      params,
      mode: 'proposed',
      drafted_by: `agent:time_sentinel@${VERSION}`,
      status: 'drafted',
      reversal_class: actionCatalog['time.self_nudge'].reversal_class,
      idempotency_key: idem,
      agent_confidence: String(result.proposal.confidence),
      agent_rationale: result.proposal.rationale,
      preview: params.message,
      originating_event_id: event.event_id,
      subject_ref: event.subject_ref,
    })
    .run();
}
